### PR TITLE
Fix more SQLAlchemy 2.0 deprecation warnings.

### PR DESCRIPTION
**Pull request by SpacemanPaul**
Still many SQLAlchemy 2.0 deprecation warnings coming out of test runs.

### Reason for this pull request

Complete work begun in #1403.

### Proposed changes

- Address outstanding SQLAlchemy 2.0 deprecation warnings in tests.
- Remove long deprecated `datacube.storage.masking` API (moved to `datacube.utils.masking around v1.8)

 - [x] Tests added / passed
 - [x] Fully documented, including `docs/about/whats_new.rst` for all changes

----
:books: Documentation preview :books:: https://datacube-core--1407.org.readthedocs.build/en/1407/

### CODE REVIEW

File: datacube/drivers/postgis/_api.py [hunk 1]
```diff
@@ -1158,20 +1158,18 @@ def list_users(self):
             inner join pg_roles user_role on am.member = user_role.oid
             where (group_role.rolname like 'odc_%%') and not (user_role.rolname like 'odc_%%')
             order by group_role.oid asc, user_role.oid asc;
-        """)
+        """))
         for row in result:
-            yield _core.from_pg_role(row['role_name']), row['user_name'], row['description']
+            yield _core.from_pg_role(row._mapping['role_name']), row._mapping['user_name'], row._mapping['description']
```
Comment (omad) [inline datacube/drivers/postgis/_api.py]: Accessing `._mapping` like this is really ugly.

[The SQLAlchemy Migration Guide](https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#result-rows-act-like-named-tuples) gives a couple of better options. Accessing these fields like `row` is a named tuple, or using `for row in result.mapping()`

Here's a quick test against SQLAlchemy 1.4.44 showing the named tuple style access works both with and without the `future=True` flag when creating the engine, and also that it works fine with raw SQL Text queries.

image

File: datacube/drivers/postgis/_core.py [hunk 2]
```diff
@@ -87,53 +87,53 @@ def ensure_db(engine, with_permissions=True):
...
-        """.format(db=quoted_db_name))
+        c.execute(text(f"""
+        grant all on database {quoted_db_name} to odc_admin;
+        """))
 
     if not has_schema(engine):
         is_new = True
         try:
-            c.execute('begin')
+            c.execute(text('begin'))
```
Comment (omad) [inline datacube/drivers/postgis/_core.py]: I'm not sure, but, surely it's possible to use `c.begin()` and friends in this section.

It's not any worse than before though, so, not super fussed. It's just ugly.

File: datacube/drivers/postgres/_api.py [hunk 3]
```diff
@@ -946,11 +954,11 @@ def check_dynamic_fields(self, concurrently=False, rebuild_views=False, rebuild_
...
 
         for metadata_type in self.get_all_metadata_types():
-            fields = get_dataset_fields(metadata_type['definition'])
-            search_fields[metadata_type['id']] = fields
+            fields = get_dataset_fields(metadata_type._mapping['definition'])
+            search_fields[metadata_type._mapping['id']] = fields
             self._setup_metadata_type_fields(
-                metadata_type['id'],
-                metadata_type['name'],
+                metadata_type._mapping['id'],
```
Comment (omad) [inline datacube/drivers/postgres/_api.py]: More SQLAlchemy results that can be treated as named tuples.

**Review by omad — CHANGES_REQUESTED** [review]

File: datacube/drivers/postgis/_core.py (hunk 2 above)
Comment (SpacemanPaul) [reply datacube/drivers/postgis/_core.py]: Yeah sure, I was just being minimal.

File: datacube/drivers/postgis/_api.py (hunk 1 above)
Comment (SpacemanPaul) [reply datacube/drivers/postgis/_api.py]: Agreed.  Updated throughout.

File: datacube/drivers/postgis/_core.py (hunk 2 above)
Comment (SpacemanPaul) [reply datacube/drivers/postgis/_core.py]: Done.

**Review by omad — APPROVED** [review]
TY

**Pull request merged**